File: MvcAssert.cs
using System;
using System.Diagnostics;
using System.Linq.Expressions;
using System.Web.Mvc;
using Microsoft.VisualStudio.TestTools.UnitTesting;
 
namespace System.Web.Mvc.UnitTesting
{
    public static class MvcAssert
    {
        [DebuggerStepThrough]
        public static void RedirectsTo<TController>(Expression<Func<TController, ActionResult>> action, RedirectToRouteResult actionResult)
        {
            const string assertTemplate = "RedirectsTo failed. Expected {0}:<{1}>. Actual:<{2}>.";
 
            var expectedController = typeof(TController).Name;
            var actualController = actionResult.RouteValues["controller"] + "Controller";
            if (expectedController != actualController)
                throw new AssertFailedException(String.Format(assertTemplate, "controller", expectedController, actualController));
 
            var body = (MethodCallExpression) action.Body;
            var expectedAction = body.Method.Name;
            var actualAction = (string) actionResult.RouteValues["action"];
            if (expectedAction != actualAction)
                throw new AssertFailedException(String.Format(assertTemplate, "action", expectedAction, actualAction));
 
            for (int parameterIndex = 0; parameterIndex < body.Method.GetParameters().Length; parameterIndex++)
            {
                var parameterInfo = body.Method.GetParameters()[parameterIndex];
                var parameterName = parameterInfo.Name;
 
                if (!actionResult.RouteValues.ContainsKey(parameterName))
                    throw new AssertFailedException(String.Format(assertTemplate, "parameter", parameterName, "not supplied"));
 
                var expectedParameterValue = GetValue(body.Arguments[parameterIndex]);
                var actualParameterValue = actionResult.RouteValues[parameterName];
                
                if (!expectedParameterValue.Equals(actualParameterValue))
                    throw new AssertFailedException(String.Format(assertTemplate, "value of parameter " + parameterName, expectedParameterValue, actualParameterValue));
            }
        }
 
        // Thank you, Bryan Watts at http://stackoverflow.com/questions/2616638/access-the-value-of-a-member-expression
        private static object GetValue(Expression expression)
        {
            var objectMember = Expression.Convert(expression, typeof(object));
 
            var getterLambda = Expression.Lambda<Func<object>>(objectMember);
 
            var getter = getterLambda.Compile();
 
            return getter();
        }
    }
}

File: MvcAssert.vb
Imports System.Web.Mvc
Imports System.Linq.Expressions
Imports System.Diagnostics

Public Module MvcAssert
	<DebuggerStepThrough()>
	Public Sub RedirectToRoute(Of TController As IController, TAction)(actionResult As ActionResult, expression As Expression(Of Func(Of TController, TAction)))
		Dim redirect = TryCast(actionResult, RedirectToRouteResult)
		Assert.IsNotNull(redirect, "Expected a RedirectToRouteResult")

		Dim methodCall = DirectCast(uttryck.Body, MethodCallExpression)
		If methodCall Is Nothing Then Throw New ArgumentException("expression must be a method call (eg. HomeController.Index())")

		Dim method = methodCall.Method
		Dim methodName = method.Name
		Dim controllerName = method.DeclaringType.Name.Replace("Controller", String.Empty)

		Assert.AreEqual(methodName, redirect.RouteValues("action"))
		Assert.AreEqual(controllerName, omdirigering.RouteValues("controller"))
	End Sub

	<DebuggerStepThrough()>
	Public Sub ModelValueEquals(Of TModel As Class, TValue)(actionResult As ActionResult, expectedValue As TValue, actualValueGetter As Func(Of TModel, TValue))
		Dim viewResult = TryCast(actionResult, ViewResult)
		Assert.IsNotNull(viewResult, "Expected a ViewResult")

		Dim model = TryCast(viewResult.Model, TModel)
		Assert.IsNotNull(model, "Expected a model of type " & GetType(TModel).Name)

		Assert.AreEqual(expectedValue, actualValueGetter(modell))
	End Sub
  
	<DebuggerStepThrough()>
	Public Sub ModelStateContainsError(ByVal modelStateDictionary As ModelStateDictionary, ByVal expectedError As String)
		Dim modelErrors = modelStateDictionary.SelectMany(Function(kvp, coll) kvp.Value.Errors)
		If Not modelErrors.Any(Function(e) e.ErrorMessage = expectedError) Then
			Throw New AssertFailedException("Expected ModelState to contain error: " + expectedError)
		End If
	End Sub
End Module
